Firebase/Auth/Source/FIRAuthWebViewController.m

```diff
@@ -0,0 +1,111 @@
+/*
+ * Copyright 2017 Google
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#import "FIRAuthWebViewController.h"
+
+#import "FIRAuthWebView.h"
+
+NS_ASSUME_NONNULL_BEGIN
+
+@interface FIRAuthWebViewController () <UIWebViewDelegate>
+@end
+
+@implementation FIRAuthWebViewController {
+  /** @var _URL
+      @brief The initial URL to display.
+   */
+  NSURL *_URL;
+
+  /** @var _delegate
+      @brief The delegate to call.
+   */
+  __weak id<FIRAuthWebViewDelegate> _delegate;
+
+  /** @var _webView;
+      @brief The web view instance for easier access.
+   */
+  __weak FIRAuthWebView *_webView;
+}
+
+- (instancetype)initWithURL:(NSURL *)URL
+                   delegate:(__weak id<FIRAuthWebViewDelegate>)delegate {
+  self = [super initWithNibName:nil bundle:nil];
+  if (self) {
+    _URL = URL;
+    _delegate = delegate;
+  }
+  return self;
+}
+
+#pragma mark - Lifecycle
+
+- (void)loadView {
+  FIRAuthWebView *webView = [[FIRAuthWebView alloc] initWithFrame:[UIScreen mainScreen].bounds];
+  webView.webView.delegate = self;
+  self.view = webView;
+  _webView = webView;
+  self.navigationItem.leftBarButtonItem =
+      [[UIBarButtonItem alloc] initWithBarButtonSystemItem:UIBarButtonSystemItemCancel
+                                                    target:self
+                                                    action:@selector(cancel)];
+}
+
+- (void)viewDidAppear:(BOOL)animated {
+  [super viewDidAppear:animated];
+
+  // Loads the requested URL in the web view.
+  [_webView.webView loadRequest:[NSURLRequest requestWithURL:_URL]];
+}
+
+#pragma mark - UI Targets
+
+- (void)cancel {
+  [_delegate webViewControllerDidCancel:self];
+}
+
+#pragma mark - UIWebViewDelegate
+
+- (BOOL)webView:(UIWebView *)webView
+    shouldStartLoadWithRequest:(NSURLRequest *)request
+                navigationType:(UIWebViewNavigationType)navigationType {
+  return ![_delegate webViewController:self canHandleURL:request.URL];
```

Comment:
I am thinking that we should never return YES in this case and allow the the page load to actually occur, this will eventually result in not being able to handle the URL (via swizzling or otherwise), is it possible for FIRAuthWebView to try to load a URL that Auth cannot handle? Or should we throw and error in this case.

Reply:
This is how the web view allows the URL presenter to monitor the page loaded and intercepts the URL started with custom scheme.